akka-stream/src/main/scala/akka/stream/stage/Stage.scala

```diff
@@ -37,6 +39,143 @@ private[stream] object AbstractStage {
+    final override def materializer: Materializer = interpreter.materializer
+
+    // FIXME: Can be removed from here and from Context
+    override private[stream] def enter(): Unit = ???
+    override private[stream] def execute(): Unit = ???
+
+    private def ctx: AsyncContext[Out, Ext] = this
+
+    private var absorbed = false
+
+    {
+      // No need to refer to the handle in a private val
+      val handler = new InHandler with OutHandler {
+        override def onPush(): Unit = stage.onPush(grab(shape.inlet), ctx)
+
+        override def onPull(): Unit = stage.onPull(ctx)
+
+        override def onUpstreamFinish(): Unit = stage.onUpstreamFinish(ctx)
+
+        override def onUpstreamFailure(ex: Throwable): Unit = stage.onUpstreamFailure(ex, ctx)
+
+        override def onDownstreamFinish(): Unit = stage.onDownstreamFinish(ctx)
+      }
+
+      setHandler(shape.inlet, handler)
+      setHandler(shape.outlet, handler)
+    }
+
+    override protected[stream] def beforePreStart(): Unit = {
+      super.beforePreStart()
+      if (stage.isDetached) pull(shape.inlet)
+    }
+
+    final override def push(elem: Out): DownstreamDirective = {
+      push(shape.outlet, elem)
+      null
+    }
+
+    final override def pull(): UpstreamDirective = {
+      pull(shape.inlet)
+      null
+    }
+
+    final override def finish(): FreeDirective = {
+      completeStage()
+      null
+    }
+
+    final override def pushAndFinish(elem: Out): DownstreamDirective = {
+      push(shape.outlet, elem)
+      completeStage()
+      null
+    }
+
+    final override def fail(cause: Throwable): FreeDirective = {
+      failStage(cause)
+      null
+    }
+
+    final override def isFinishing: Boolean = isClosed(shape.inlet)
+
+    final override def absorbTermination(): TerminationDirective = {
+      if (isClosed(shape.outlet)) {
+        val ex = new UnsupportedOperationException("It is not allowed to call absorbTermination() from onDownstreamFinish.")
+        // This MUST be logged here, since the downstream has cancelled, i.e. there is noone to send onError to, the
+        // stage is just about to finish so noone will catch it anyway just the interpreter
+
+        interpreter.log.error(ex.getMessage)
+        throw ex // We still throw for correctness (although a finish() would also work here)
+      }
+      if (isAvailable(shape.outlet)) stage.onPull(ctx)
+      null
+    }
+
+    override def pushAndPull(elem: Out): FreeDirective = {
+      push(shape.outlet, elem)
+      pull(shape.inlet)
+      null
+    }
+
+    final override def holdUpstreamAndPush(elem: Out): UpstreamDirective = {
+      push(shape.outlet, elem)
+      null
+    }
+
+    final override def holdDownstreamAndPull(): DownstreamDirective = {
+      pull(shape.inlet)
+      null
+    }
+
+    final override def isHoldingDownstream: Boolean = isAvailable(shape.outlet)
+
+    final override def isHoldingUpstream: Boolean = !(isClosed(shape.inlet) || hasBeenPulled(shape.inlet))
```

Comment:
emulating the old behavior wastes space, so unless it has real usage advantages I’d go with the current implementation